src/shims/panic.rs

```diff
@@ -0,0 +1,180 @@
+use rustc::mir;
+use crate::*;
+use super::machine::FrameData;
+use rustc_target::spec::PanicStrategy;
+use crate::rustc_target::abi::LayoutOf;
+
+/// Holds all of the relevant data for a call to
+/// __rust_maybe_catch_panic
+///
+/// If a panic occurs, we update this data with
+/// the information from the panic site
+#[derive(Debug)]
+pub struct CatchUnwindData<'tcx> {
+    /// The 'data' argument passed to `__rust_maybe_catch_panic`
+    pub data: Pointer<Tag>,
+    /// The `data_ptr` argument passed to `__rust_maybe_catch_panic`
+    pub data_ptr: MPlaceTy<'tcx, Tag>,
+    /// The `vtable_ptr` argument passed to `__rust_maybe_catch_panic`
+    pub vtable_ptr: MPlaceTy<'tcx, Tag>,
+    /// The `dest` from the original call to `__rust_maybe_catch_panic`
+    pub dest: PlaceTy<'tcx, Tag>,
+    /// The `ret` from the original call to `__rust_maybe_catch_panic`
+    pub ret: mir::BasicBlock,
+}
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+
+    /// Handles the special "miri_start_panic" intrinsic, which is called
+    /// by libpanic_unwind to delegate the actual unwinding process to Miri
+    #[inline(always)]
+    fn handle_miri_start_panic(
+        &mut self,
+        args: &[OpTy<'tcx, Tag>],
+        unwind: Option<mir::BasicBlock>
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        trace!("miri_start_panic: {:?}", this.frame().span);
+
+        if this.tcx.tcx.sess.panic_strategy() == PanicStrategy::Abort  {
+            // FIXME: Actually print out the payload here
+            throw_unsup_format!("the evaluated program panicked");
+        }
+
+        // Get the raw pointer stored in arg[0]
+        let scalar = this.read_immediate(args[0])?;
+        this.machine.panic_payload = Some(scalar);
+
+        // Jump to the unwind block to begin unwinding
+        // We don't use 'goto_block' - if `unwind` is None,
+        // we'll end up immediately returning out of the
+        // function during the next step() call
+        let next_frame = this.frame_mut();
+        next_frame.block = unwind;
+        next_frame.stmt = 0;
+        return Ok(())
+    }
+
+    #[inline(always)]
+    fn handle_catch_panic(
+        &mut self,
+        args: &[OpTy<'tcx, Tag>],
+        dest: PlaceTy<'tcx, Tag>,
+        ret: mir::BasicBlock,
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+        let tcx = &{this.tcx.tcx};
+
+        // fn __rust_maybe_catch_panic(
+        //     f: fn(*mut u8),
+        //     data: *mut u8,
+        //     data_ptr: *mut usize,
+        //     vtable_ptr: *mut usize,
+        // ) -> u32
+        let f = this.read_scalar(args[0])?.not_undef()?;
+        let data = this.read_scalar(args[1])?.not_undef()?;
+        let data_ptr = this.deref_operand(args[2])?;
+        let vtable_ptr = this.deref_operand(args[3])?;
```

Comment:
These are actually places, right? You deref'd them. So `_place` would be a more accurate suffix than `_ptr`.

Reply:
They logically represent the pointers passed by the caller, so I thought calling them `_ptr` made things clearer.

Reply:
Given that we have `Pointer` (rvalue) and `Place` (lvalue) with different APIs, I think we should be more precise here.